src/main/kotlin/solve/rendering/engine/scene/Scene.kt

```diff
@@ -1,39 +1,58 @@
 package solve.rendering.engine.scene
 
-import solve.rendering.engine.core.renderers.Renderer
-
-class Scene(renderers: List<Renderer> = emptyList()) {
-    private val _renderObjects = mutableListOf<RenderObject>()
-    val renderObjects: List<RenderObject>
-        get() = _renderObjects
-
-    private val _renderers = renderers
-    val renderers: List<Renderer>
-        get() = _renderers
-
-    fun update(deltaTime: Float) {
-        _renderObjects.forEach { gameObject ->
-            if (gameObject.isDestroyed) {
-                removeRenderObject(gameObject)
-                return@forEach
-            }
-
-            gameObject.update(deltaTime)
-            _renderers.forEach { it.render() }
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import solve.rendering.engine.core.renderers.FramesRenderer
+import solve.rendering.engine.core.renderers.LandmarkLayerRenderer
+import solve.scene.model.Layer
+import java.util.concurrent.CopyOnWriteArrayList
+
+class Scene(val framesRenderer: FramesRenderer) {
+    private val _landmarkRenderers = CopyOnWriteArrayList<LandmarkLayerRenderer>()
+    private val _landmarkRenderersLayers = mutableMapOf<LandmarkLayerRenderer, Layer>()
+    val landmarkRenderers: List<LandmarkLayerRenderer>
+        get() = _landmarkRenderers
+    val landmarkLayerRendererLayers: Map<LandmarkLayerRenderer, Layer>
+        get() = _landmarkRenderersLayers
+
+    var isFramesRendererInitializing = false
+
+    fun update() {
+        render()
+    }
+
+    fun addLandmarkRenderer(renderer: LandmarkLayerRenderer, layer: Layer) {
+        _landmarkRenderers.add(renderer)
+        _landmarkRenderersLayers[renderer] = layer
+    }
+
+    fun clearLandmarkRenderers() {
+        landmarkRenderers.forEach { it.delete() }
+        _landmarkRenderers.clear()
+        _landmarkRenderersLayers.clear()
+    }
+
+    fun initializeFramesRenderer() {
+        CoroutineScope(Dispatchers.Default).launch {
+            isFramesRendererInitializing = true
+            delay(FrameRendererInitializationDelayMillis)
+            isFramesRendererInitializing = false
         }
     }
 
-    fun addRenderObject(renderObject: RenderObject) {
-        if (_renderObjects.contains(renderObject)) {
-            println("The scene already contains adding game object ($renderObject)!")
+    private fun render() {
+        framesRenderer.render()
+        if (isFramesRendererInitializing) {
```

Comment:
looks like you are using it from different threads without volatile

Reply:
OpenGL renders only in one thread and this function called from rendering context thread.

Reply:
I noticed that `isFramesRendererInitializing` really needs a volatile annotation